### ISSUE TITLE
Kokkos does not compile with -DLAMMPS_BIGBIG

### ISSUE THREAD

**Issue by rbberger** [lammps/lammps#365]
Trying to compile LAMMPS with Kokkos and -DLAMMPS_BIGBIG currently fails.

It turns out to be caused by the `molecule` array. It is declared as regular 1D `int` array, however, it's actually a 1D `tagint` array. PR incoming...
```
mpicxx -g -O3  -I/usr/include/hdf5/serial -DFFT_KISSFFT -DLAMMPS_GZIP -DLAMMPS_PNG -DLAMMPS_JPEG -DLAMMPS_BIGBIG -Wall -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-maybe-uninitialized -DLMP_KOKKOS  -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX=1   -I./ -I../../lib/kokkos/core/src -I../../lib/kokkos/containers/src -I../../lib/kokkos/algorithms/src  --std=c++11 -fopenmp  -c ../npair_kokkos.cpp
../npair_kokkos.cpp: In instantiation of 'void LAMMPS_NS::NPairKokkos<DeviceType, HALF_NEIGH, GHOST>::build(LAMMPS_NS::NeighList*) [with DeviceType = Kokkos::OpenMP; int HALF_NEIGH = 0; int GHOST = 0]':
../npair_kokkos.cpp:739:16:   required from here
../npair_kokkos.cpp:173:37: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   if(list->d_neighbors.dimension_0()<nall) {
                                      
../npair_kokkos.cpp: In instantiation of 'void LAMMPS_NS::NPairKokkos<DeviceType, HALF_NEIGH, GHOST>::build(LAMMPS_NS::NeighList*) [with DeviceType = Kokkos::OpenMP; int HALF_NEIGH = 0; int GHOST = 1]':
../npair_kokkos.cpp:740:16:   required from here
../npair_kokkos.cpp:173:37: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
../npair_kokkos.cpp: In instantiation of 'void LAMMPS_NS::NPairKokkos<DeviceType, HALF_NEIGH, GHOST>::build(LAMMPS_NS::NeighList*) [with DeviceType = Kokkos::OpenMP; int HALF_NEIGH = 1; int GHOST = 0]':
../npair_kokkos.cpp:741:16:   required from here
../npair_kokkos.cpp:173:37: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
../npair_kokkos.cpp: In instantiation of 'void LAMMPS_NS::NPairKokkos<DeviceType, HALF_NEIGH, GHOST>::build(LAMMPS_NS::NeighList*) [with DeviceType = Kokkos::OpenMP; int HALF_NEIGH = 1; int GHOST = 1]':
../npair_kokkos.cpp:742:16:   required from here
../npair_kokkos.cpp:173:37: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
In file included from ../../lib/kokkos/core/src/Kokkos_Parallel.hpp:52:0,
                 from ../../lib/kokkos/core/src/Kokkos_OpenMP.hpp:60,
                 from ../../lib/kokkos/core/src/Kokkos_Core.hpp:57,
                 from ../kokkos_type.h:19,
                 from ../neigh_list_kokkos.h:19,
                 from ../npair_kokkos.h:62,
                 from ../npair_kokkos.cpp:14:
../../lib/kokkos/core/src/Kokkos_View.hpp: In instantiation of 'Kokkos::View<DataType, Properties>::View(const Kokkos::View<RT, RP ...>&) [with RT = long int*; RP = {Kokkos::LayoutRight, Kokkos::OpenMP, void}; DataType = const int*; Properties = {Kokkos::LayoutRight, Kokkos::OpenMP, void}]':
../npair_kokkos.cpp:152:63:   required from 'void LAMMPS_NS::NPairKokkos<DeviceType, HALF_NEIGH, GHOST>::build(LAMMPS_NS::NeighList*) [with DeviceType = Kokkos::OpenMP; int HALF_NEIGH = 0; int GHOST = 0]'
../npair_kokkos.cpp:739:16:   required from here
../../lib/kokkos/core/src/Kokkos_View.hpp:1221:7: error: static assertion failed: Incompatible View copy construction
       static_assert( Mapping::is_assignable , "Incompatible View copy construction" );
       ^~~~~~~~~~~~~
In file included from ../npair_kokkos.cpp:14:0:
../npair_kokkos.h: In instantiation of 'LAMMPS_NS::NeighborKokkosExecute<DeviceType>::NeighborKokkosExecute(const LAMMPS_NS::NeighListKokkos<DeviceType>&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_xfloat_2d_randomread&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_2d&, int, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_3&, int, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_x_array_randomread&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_const&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_const&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_const&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_tagint_1d_const&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_tagint_2d_const&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_2d_const&, const int&, const int&, const int&, const int&, const int&, const int&, const int&, const int&, const int&, const int&, const X_FLOAT&, const X_FLOAT&, const X_FLOAT&, const int&, const int&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_const&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_const&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_2d_const&, const int&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_const&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_const&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_const&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_const&, const int&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_const&, const typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_const&, const X_FLOAT*, const X_FLOAT*, const int&, const int&, const int&, const int&, const int&, const int&) [with DeviceType = Kokkos::OpenMP; typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_xfloat_2d_randomread = Kokkos::View<const double**, Kokkos::LayoutRight, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<2u> >; typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d = Kokkos::View<int*, Kokkos::LayoutRight, Kokkos::OpenMP, void>; typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_2d = Kokkos::View<int**, Kokkos::LayoutRight, Kokkos::OpenMP, void>; typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_3 = Kokkos::View<int* [3], Kokkos::LayoutRight, Kokkos::OpenMP, void>; typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_x_array_randomread = Kokkos::View<const double* [3], Kokkos::LayoutRight, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<2u> >; typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_1d_const = Kokkos::View<const int*, Kokkos::LayoutRight, Kokkos::OpenMP, void>; typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_tagint_1d_const = Kokkos::View<const long int*, Kokkos::LayoutRight, Kokkos::OpenMP, void>; typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_tagint_2d_const = Kokkos::View<const long int**, Kokkos::LayoutRight, Kokkos::OpenMP, void>; typename LAMMPS_NS::NeighborKokkosExecute<DeviceType>::AT::t_int_2d_const = Kokkos::View<const int**, Kokkos::LayoutRight, Kokkos::OpenMP, void>; X_FLOAT = double]':
../npair_kokkos.cpp:152:63:   required from 'void LAMMPS_NS::NPairKokkos<DeviceType, HALF_NEIGH, GHOST>::build(LAMMPS_NS::NeighList*) [with DeviceType = Kokkos::OpenMP; int HALF_NEIGH = 0; int GHOST = 0]'
../npair_kokkos.cpp:739:16:   required from here
../npair_kokkos.h:165:13: warning: 'LAMMPS_NS::NeighborKokkosExecute<Kokkos::OpenMP>::nlocal' will be initialized after [-Wreorder]
   const int nlocal;
             ^~~~~~
../npair_kokkos.h:149:43: warning:   'const t_x_array_randomread LAMMPS_NS::NeighborKokkosExecute<Kokkos::OpenMP>::x' [-Wreorder]
   const typename AT::t_x_array_randomread x;
                                           ^
../npair_kokkos.h:175:3: warning:   when initialized here [-Wreorder]
   NeighborKokkosExecute(
   ^~~~~~~~~~~~~~~~~~~~~
../npair_kokkos.h:162:33: warning: 'LAMMPS_NS::NeighborKokkosExecute<Kokkos::OpenMP>::bininvz' will be initialized after [-Wreorder]
   const X_FLOAT bininvx,bininvy,bininvz;
                                 ^~~~~~~
../npair_kokkos.h:119:13: warning:   'const int LAMMPS_NS::NeighborKokkosExecute<Kokkos::OpenMP>::exclude' [-Wreorder]
   const int exclude;
             ^~~~~~~
../npair_kokkos.h:175:3: warning:   when initialized here [-Wreorder]
   NeighborKokkosExecute(
   ^~~~~~~~~~~~~~~~~~~~~
Makefile:105: recipe for target 'npair_kokkos.o' failed
```

**Comment by stanmoore1** [comment]
@rbberger good catch.

**Comment by akohlmey** [comment]
There is another `-DLAMMPS_BIGBIG` issue in `fix_qeq_reax_kokkos.cpp` and `fix_qeq_reax_kokkos.h`, plus it uses `tag[i]` and `tag[j]` where it looks as if the code should use `itag` and `jtag`, which would be initialized from `tag(i)` and `tag(j)` instead.

**Comment by stanmoore1** [comment]
@akohlmey yes fix_qeq_reax_kokkos.cpp does have tagint issues, I'll take a look. Though I'd be surprised if anyone is running > 2 billion atom systems with ReaxFF ... :)

**Comment by akohlmey** [comment]
no worries, i have the changes already.

**Comment by stanmoore1** [comment]
Oh, OK thanks

**Comment by giacomofiorin** [comment]
@stanmoore1 2 billion atoms are certainly huge for most simulation, but there is a niche application of shock simulations (e.g. Vashishta) who do not need to simulate for very long times but need a reactive model.

**Comment by stanmoore1** [comment]
@giacomofiorin interesting...

**Comment by stanmoore1** [comment]
@akohlmey looks like there are also tagint issues with pair_reax_c_kokkos and pair_tersoff*_kokkos. I just did a quick search, could be more. Do you want to take those or should I?

**Comment by akohlmey** [comment]
my changes are in rbberger/lammps#1

**Comment by akohlmey** [comment]
@stanmoore1 they are probably hidden by explicit assignments, so the compilation succeeded (at least with OpenMP). Typically, compilers only stop when assigning pointers. Thus it is occasionally a good idea to explicitly grep for "tag" and check whether anything is assigned to an `int` or `int *` or similar.

I'm about to head home, so please knock yourself out.

**Comment by stanmoore1** [comment]
@akohlmey OK thanks for the help

**Comment by stanmoore1** [comment]
Fixed in PR #366. Closed #365.
